Logic Overhaul and Logic Tricks

AmazingAmpharos opened the pull request:
This is mostly just me resubmitting the PR with the various naming and synchronizations already known. I also fixed up a few small logic issues in Pinnacle Rock and the moon. Still probably undertested, but it runs and makes seeds that look like they are using what it's supposed to do correctly.

To repeat I guess, this PR's main purpose is twofold. It tightens up logic to generally be as precisely accurate as possible across the game (generally a lot more changes in more logically complex areas like Snowhead, a lot of the stuff in places like Clock Town was just great without changes) and implements logic tricks to allow end users to customize the logic to their personal comfort level (more experienced players can turn on tricks to make more demanding seeds; beginners can leave everything off and play the game within safer bounds).

AmazingAmpharos committed:
Fix PR to synchronize with recent work

This is mostly just me resubmitting the PR with the various naming and synchronizations already known. I also fixed up a few small logic issues in Pinnacle Rock and the moon. Still probably undertested, but it runs and makes seeds that look like they are using what it's supposed to do correctly.

PixelShake92 reviewed (commented):
I think these changes are fine. I see why you have chosen to do most of the changes and it mostly comes down to personal opinion and preference. Our original goal was to make things accessible with no tricks so I think these changes reflect the desire to move towards trick customization.

G4M3RL1F3 commented:
I spent some time today to make sure logic tricks are applied to logic and that they're properly sent to slot data, for the purpose of the tracker I'm working on. Here are some things that should be looked at:

In `Options.py`, a new `class` should be added, which should also be added in `dataclass`.
```Python
class LogicTricks(OptionSet):
    """Select tricks that may be logically required."""
    display_name = "Logic Tricks"
    valid_keys = frozenset(LOGIC_TRICKS)

@dataclass
class MMROptions(PerGameCommonOptions):
    ...
    logic_tricks: Logic Tricks
```

The way on how listing the tricks is written in `LogicTricks.py` could be simplified. Instead, a list can be made with only the names of the tricks. The only downside is that there's no description for each trick. An idea is to add a new `.md` file to serve that purpose.
```Python
LOGIC_TRICKS = [
    "Early Game Hard Combat",
    "Stock Pot Inn at Night with Nothing",
    ...
]
```

In `__init__.py`, add the logic tricks to slot data (already done in this PR and does not need to be changed).

I also tested all of this to make sure it works:
- I generated multiple seeds with the logic trick `Early Game Hard Combat` on and off to see if the Dodongo and Peahat grottos would move from sphere 1 to a later one.
  - Turning the trick off always put them in sphere 1, while turning it on never puts them in sphere 1. It confirms that the logic tricks affect logic when turned on, and doesn't when turned off.
- I hosted a seed on the AP website to see if my tracker would fetch the correct states of each logic trick. In that seed, I had one trick on, the other was off.
  - My tracker correctly displayed the states of each logic trick.

If a PR can be helpful, I can make one. Be mindful that I only included two tricks for the sake of making it work. My `NormalRules.py` file would only have locations affected by `Early Game Hard Combat` and `Stock Pot Inn at Night with Nothing`. The rest would be unchanged.

LittleCube-hax commented on worlds/mm_recomp/NormalRules.py:
```diff
@@ -157,6 +159,53 @@ def has_enough_remains(state, player, need_count):
+                       state.can_reach("Ikana Well Invisible Chest", 'Location', player)
+                   )
+               )
+           )
```
this block looks like it's indented one tab plus three spaces too far to the right

LittleCube-hax commented on worlds/mm_recomp/NormalRules.py:
```diff
@@ -157,6 +159,53 @@ def has_enough_remains(state, player, need_count):
+                logic_trick(options, "Goron Bomb Jumps") and
+                state.has("Goron Mask", player) and
+                state.has("Progressive Bomb Bag", player)
+            ))
```
same with this block, but just one tab too far to the right

LittleCube-hax commented on worlds/mm_recomp/NormalRules.py:
```diff
@@ -157,6 +159,53 @@ def has_enough_remains(state, player, need_count):
     return remains_count >= need_count
 
+def logic_trick(options, trick_name):
+    return (trick_name in options.logic_tricks.value)
```
bear with me since this is an insane nitpick, but i think this should be a flag bitmask or an array of bools instead, just to play it extra safe for gen speed

i'm willing to implement this myself since it's such a crazy thing to ask lol

LittleCube-hax replied:
resolving since this is not required to merge at all lmao

LittleCube-hax commented on worlds/mm_recomp/NormalRules.py:
```diff
@@ -167,18 +216,33 @@ def get_region_rules(player, options):
+                        state.has("Deku Mask", player) or
+                        (
+                            logic_trick(options, "Cross Poisoned Swamp as Goron or Zora") and
+                            (
+                                state.has("Goron Mask", player) or
+                                state.has("Zora Mask", player)
+                            )
+                        )
+                    )
                 )
             ),
         "Southern Swamp (Deku Palace) -> Swamp Spider House":
-            lambda state: state.has("Deku Mask", player),
+            lambda state: (
+                state.has("Deku Mask", player) or
+                (
+                    logic_trick(options, "Cross Poisoned Swamp as Goron or Zora") and
+                    (
+                        state.has("Goron Mask", player) or
+                        state.has("Zora Mask", player)
+                    )
```
since i see a bunch of these `logic_trick` blocks that are carbon copies of each other, i think we should have each trick as its own macro and call that from the lambda (at the very least the ones that are more than one line and duplicated) which will also make maintaining tricks easier down the road

LittleCube-hax commented on worlds/mm_recomp/NormalRules.py:
```diff
@@ -969,70 +1127,68 @@ def get_location_rules(player, options):
+        # boss without a real melee weapon is completely unfair to the player.
         "Woodfall Temple Heart Container":
             lambda state: (
-                can_smack(state, player) and 
```
this is an example of an obscure method, something i anticipated a while back: this phenomenon keeps popping up where some players prefer not to use some of the methods required by logic due to required obscure knowledge or having to do something cumbersome/out of the way, but the method to do it is not considered a trick or a glitch

i think we should have another set of togglables `obscure_methods` or a similar name, and relegate these kinds of logical methods to this set that can be enabled by players who want them

LittleCube-hax replied:
resolving since this is not required to merge, we'll probably handle this later

LittleCube-hax reviewed (changes requested):
first of all, just wanted to say thank you for doing all of this ampharos, really helpful stuff

i went ahead and commented on what i noticed, if you like you can certainly tackle any of it that you want but there's definitely some stuff i noted that we'll probably just do ourselves later lol

ThatHypedPerson commented:
We also noticed that the `Options.py` file was missing from this PR.

AmazingAmpharos committed:
Add missing options file

Accidentally excluded this file adding the logic tricks to the options file. Rectifying.

AmazingAmpharos committed:
Fix descriptions

A few descriptions were inaccurate or lacking detail. This addresses that.

AmazingAmpharos committed:
Functionalize common trick logic

This functionalizes the common calls to the tricks "Goron Bomb Jumps", "Fewer Lens Requirements", "Powder Keg as Explosive", and "Climb Ikana Canyon without Ice Arrows". This results in considerably more terse logic. All functions I wrote are also, in theory, synchronized to the Allman's formatting style. I did not address functions that were pre-existing for style revision.

ThatHypedPerson committed:
Merge latest changes into PR

ThatHypedPerson committed:
Merge branch 'tricks' into 0.9.5
